turnstile-example/turnstile/examples/cs478/current/stlc.rkt

```diff
@@ -442,3 +446,23 @@
   #;  [⊢ (if- (inl-? e-)
           (let- ([x- (inl--val e-)]) e1-)
           (let- ([y- (inr--val e-)]) e2-)) ⇒ τ])
+
+(provide typed-match)
+(define-typerule typed-match
+  [(_ (x:id e) b) ≫
+   [⊢ e ≫ e- ⇒ τ]
+   [[x ≫ x- : τ] ⊢ b ≫ b- ⇒ τ-out]
+   ---------------------------------
+   [⊢ (let- ([x- e-]) b-) ⇒ τ-out]]
+  [(_ ([x:id ...] e) b) ≫
+   [⊢ e ≫ e- ⇒ (~× τ-tup ...)]
+   #:fail-unless (= (stx-length #'(x ...)) (stx-length #'(τ-tup ...)))
+   (format "~a pattern variables does not match ~a values in tuple" (stx-length #'(x ...)) (stx-length #'(τ-tup ...)))
+   [[x ≫ x- : τ-tup] ... ⊢ b ≫ b- ⇒ τ-out]
+   -----------------------------------------
+   [⊢ (match- e- [(list- x- ...) b-]) ⇒ τ-out]]
```

Comment:
I wasn't sure what the best conclusions to this and the following type rules are. It overlaps partly with previous type rules and helper functions, but not completely.

Reply:
The idea was that, to match a tuple, you would write:
```(typed-match [(a b c d) (tup 1 2 3 4)] a) -> 1```
and for records you would do:
```(typed-match [(label-1 label-2) (rec [label-1 = 3] [label-2 = 4])] (+ label-1 label-2)) -> 7```
In other words, you would specify the labels of the record and they would become identifiers containing the values corresponding to that label in the record.

Reply:
Ok I get it. It's a little non-standard because the label doubles as the binder. The book has separate labels and binders right?

Reply:
Yep, I see that now, gonna change it to the format used in the book.

Reply:
Is there a way to interleave typed and untyped conclusions? I.e. I want to extend `typed-match` to support multiple matches at a time (like `match-let`):
```
(typed-match
    ([pat1 e1]
     [pat2 e2]
     ...)
     body)
```

But I'm not sure what the conclusion should look like, because I want to translate it into matching the first pattern and then expanding the rest in a new typed-match form.

Reply:
It depends on the nesting. 

You can embed typed terms in untyped terms and the typed part will be checked normally while the untyped part will just ignore the types.

The reverse way wont work though, because the typed term will error when it cant find the type information.

Reply:
The new syntax looks like:

```
(typed-match ([([label1 = (a b)]) (rec [label1 = (tup 1 #f)] [label2 = 5])]) b) -> #f
```